contrib/devtools/xversionkeys.py

```diff
@@ -97,6 +100,20 @@ def writeTableCPPH(outfile, table):
         print ("    {%40s, %42s }," % (x["name"], "xvt_"+x["valtype"]), file = outfile)
     print("}; // const unordered_map valtype\n\n\n", file = outfile)
 
+    # set key types in enum
+    print("enum {", file = outfile)
+    valtypes = set(x["keytype"] for x in table.values())
+    for x in sorted(valtypes):
+        print ("    %20s," % ("xkt_"+x), file = outfile)
+    print("}; // enum keytypes\n\n\n", file = outfile)
+
+    # map keys to their expected value type
+    print("const std::unordered_map<uint64_t, std::string> keytype = {", file = outfile)
+    for k in sorted(table.keys()):
+        x = table[k]
+        print ("    {%40s, %42s }," % (x["name"], '"'+x["keytype"]+'"'), file = outfile)
```

Comment:
I strongly suggest to use the `enum` values in the table and make it an unordered map `uint64_t` -> that `enum`.

This makes the code using it "type safer", in the sense that it should be relatively easy to later on make the keytype a more typesafe enum and go with that. At least no more risk to make a simple typo and have the code fail. Also (though a lot less important), strings are probably a tad slower to handle in comparisons etc.

Reply:
i also agree with this. i have made that change